pandas/core/arrays/masked.py

```diff
@@ -0,0 +1,117 @@
+from typing import TYPE_CHECKING
+
+import numpy as np
+
+from pandas.core.dtypes.common import is_integer
+from pandas.core.dtypes.missing import isna, notna
+
+from pandas.core.algorithms import take
+from pandas.core.arrays import ExtensionArray, ExtensionOpsMixin
+import pandas.core.common as com
+from pandas.core.indexers import check_bool_array_indexer
+
+if TYPE_CHECKING:
+    from pandas._typing import Scalar
+
+
+class BaseMaskedArray(ExtensionArray, ExtensionOpsMixin):
+    """
+    Base class for masked arrays (which use _data and _mask to store the data).
+
+    numpy based
+    """
+
+    _data: np.ndarray
+    _mask: np.ndarray
+
+    # The value used to fill '_data' to avoid upcasting
+    _internal_fill_value: "Scalar"
+
+    def __getitem__(self, item):
+        if is_integer(item):
+            if self._mask[item]:
+                return self.dtype.na_value
+            return self._data[item]
+
+        elif com.is_bool_indexer(item):
+            item = check_bool_array_indexer(self, item)
+
+        return type(self)(self._data[item], self._mask[item])
+
+    def __iter__(self):
+        for i in range(len(self)):
+            if self._mask[i]:
+                yield self.dtype.na_value
+            else:
+                yield self._data[i]
+
+    def __len__(self) -> int:
+        return len(self._data)
+
+    __array_priority__ = 1000  # higher than ndarray so ops dispatch to us
+
+    def __array__(self, dtype=None):
+        """
+        the array interface, return my values
+        We return an object array here to preserve our scalar values
+        """
+        return self.to_numpy(dtype=dtype)
+
+    def __arrow_array__(self, type=None):
+        """
+        Convert myself into a pyarrow Array.
+        """
+        import pyarrow as pa
+
+        return pa.array(self._data, mask=self._mask, type=type)
+
+    @property
+    def _hasna(self) -> bool:
```

Comment:
I suspect IntegerArray has some `isna().any()` that can be updated to use this property.